File: FDChessBoardView/FDChessboardView.m
//
//  FDChessboardView.m
//  FDChessboardView
//
//  Created by William Entriken on 9/12/14.
//  Copyright (c) 2014 William Entriken. All rights reserved.
//

#import "FDChessboardView.h"
#import <UIImage+SVG/UIImage+SVG.h>

@interface FDChessboardView()
// These arrays use index i=0..63 from white queen rook to king rook, then ranks up to black pieces
@property (strong, nonatomic) NSMutableArray *tilesPerSquare;
@property (strong, nonatomic) NSMutableArray *piecesPerSquare;
@property (strong, nonatomic) NSMutableArray *pieceImageViewsPerSquare;
@property (strong, nonatomic) UIView *lastMoveArrow;
@property (strong, nonatomic) UIView *premoveArrow;
@property (strong, nonatomic) UIColor *lightPieceColor; // temporary until SVG is fixed
@property (strong, nonatomic) UIColor *darkPieceColor; // temporary until SVG is fixed
@end

@implementation FDChessboardView

#pragma mark - FDChessboardView properties

- (UIColor *)lightBackgroundColor
{
    if (!_lightBackgroundColor)
        _lightBackgroundColor = [UIColor colorWithRed:222.0/255 green:196.0/255 blue:160.0/255 alpha:1];
    return _lightBackgroundColor;
}

- (UIColor *)darkBackgroundColor
{
    if (!_darkBackgroundColor)
        _darkBackgroundColor = [UIColor colorWithRed:160.0/255 green:120.0/255 blue:55.0/255 alpha:1];
    return _darkBackgroundColor;
}

- (UIColor *)targetBackgroundColor
{
    if (!_targetBackgroundColor)
        _targetBackgroundColor = [UIColor colorWithHue:0.75 saturation:0.5 brightness:0.5 alpha:1.0];
    return _targetBackgroundColor;
}

- (UIColor *)legalBackgroundColor
{
    if (!_legalBackgroundColor)
        _legalBackgroundColor = [UIColor colorWithHue:0.25 saturation:0.5 brightness:0.5 alpha:1.0];
    return _legalBackgroundColor;
}

- (UIColor *)lightPieceColor
{
    if (!_lightPieceColor)
        _lightPieceColor = [UIColor whiteColor];
    return _lightPieceColor;
}

- (UIColor *)darkPieceColor
{
    if (!_darkPieceColor)
        _darkPieceColor = [UIColor blackColor];
    return _darkPieceColor;
}

- (UIColor *)lastMoveColor
{
    if (!_lastMoveColor)
        _lastMoveColor = [UIColor colorWithHue:0.35 saturation:0.5 brightness:0.5 alpha:1.0];
    return _lastMoveColor;
}

- (UIColor *)premoveColor
{
    if (!_premoveColor)
        _premoveColor = [UIColor colorWithHue:0.15 saturation:0.5 brightness:0.5 alpha:1.0];
    return _premoveColor;
}

- (void)setDataSource:(id<FDChessboardViewDataSource>)dataSource
{
    _dataSource = dataSource;
    [self reloadData];
}

#pragma mark - FDChessboardView private properties

- (NSMutableArray *)tilesPerSquare
{
    if (!_tilesPerSquare) {
        _tilesPerSquare = [NSMutableArray array];
        [self resetConstraintsAndSetTiles];
    }
    return _tilesPerSquare;
}

- (NSMutableArray *)piecesPerSquare
{
    if (!_piecesPerSquare) {
        _piecesPerSquare = [NSMutableArray array];
    }
    return _piecesPerSquare;
}

- (NSMutableArray *)pieceImageViewsPerSquare
{
    if (!_pieceImageViewsPerSquare) {
        _pieceImageViewsPerSquare = [NSMutableArray array];
    }
    return _pieceImageViewsPerSquare;
}

#pragma mark - UIView Initializers

- (void)setup
{
    [self resetConstraintsAndSetTiles];
//    [self insertPiece];
}

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    [self setup];
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    [self setup];
    return self;
}

/*
 
- (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event
{
    UITouch *touch = [touches anyObject];
    CGPoint point = [touch locationInView:self];
    NSInteger x = point.x*8/self.frame.size.width;
    NSInteger y = 8-point.y*8/self.frame.size.height;
    UIView *tile = self.tilesPerSquare[y*8+x];
    tile.backgroundColor = self.targetBackgroundColor;
}

- (void)touchesMoved:(NSSet *)touches withEvent:(UIEvent *)event
{
    UITouch *touch = [touches anyObject];
    CGPoint point = [touch locationInView:self];
    if (![self pointInside:point withEvent:event])
        return;
    NSInteger x = point.x*8/self.frame.size.width;
    NSInteger y = 8-point.y*8/self.frame.size.height;
    UIView *tile = self.tilesPerSquare[y*8+x];
    tile.backgroundColor = self.legalBackgroundColor;
}
 
 */

#pragma mark - Drawing helpers

- (void)resetConstraintsAndSetTiles
{
    self.translatesAutoresizingMaskIntoConstraints = NO;
    self.tilesPerSquare = [NSMutableArray array];
    for (int i=0; i<64; i++) {
        UIView *view = [[UIView alloc] init];
        view.translatesAutoresizingMaskIntoConstraints = NO;
        view.tag = i;
        view.backgroundColor = !(i%2) ^ ((i/8)%2) ? self.darkBackgroundColor : self.lightBackgroundColor;
        [self addSubview:view];
        [self.tilesPerSquare addObject:view];
    }
    for (int i=0; i<64; i++) {
        NSLayoutConstraint *constraint;
        if (i<8) {
            constraint = [NSLayoutConstraint constraintWithItem:self.tilesPerSquare[i]
                                                      attribute:NSLayoutAttributeBottom
                                                      relatedBy:NSLayoutRelationEqual
                                                         toItem:self
                                                      attribute:NSLayoutAttributeBottom
                                                     multiplier:1
                                                       constant:0];
            [self addConstraint:constraint];
        } else {
            constraint = [NSLayoutConstraint constraintWithItem:self.tilesPerSquare[i]
                                                      attribute:NSLayoutAttributeBottom
                                                      relatedBy:NSLayoutRelationEqual
                                                         toItem:self.tilesPerSquare[i-8]
                                                      attribute:NSLayoutAttributeTop
                                                     multiplier:1
                                                       constant:0];
            [self addConstraint:constraint];
            constraint = [NSLayoutConstraint constraintWithItem:self.tilesPerSquare[i]
                                                      attribute:NSLayoutAttributeHeight
                                                      relatedBy:NSLayoutRelationEqual
                                                         toItem:self.tilesPerSquare[i-8]
                                                      attribute:NSLayoutAttributeHeight
                                                     multiplier:1
                                                       constant:0];
            [self addConstraint:constraint];
        }
        if (i%8==0) {
            constraint = [NSLayoutConstraint constraintWithItem:self.tilesPerSquare[i]
                                                      attribute:NSLayoutAttributeLeft
                                                      relatedBy:NSLayoutRelationEqual
                                                         toItem:self
                                                      attribute:NSLayoutAttributeLeft
                                                     multiplier:1
                                                       constant:0];
            [self addConstraint:constraint];
        } else {
            constraint = [NSLayoutConstraint constraintWithItem:self.tilesPerSquare[i]
                                                      attribute:NSLayoutAttributeLeft
                                                      relatedBy:NSLayoutRelationEqual
                                                         toItem:self.tilesPerSquare[i-1]
                                                      attribute:NSLayoutAttributeRight
                                                     multiplier:1
                                                       constant:0];
            [self addConstraint:constraint];
            constraint = [NSLayoutConstraint constraintWithItem:self.tilesPerSquare[i]
                                                      attribute:NSLayoutAttributeWidth
                                                      relatedBy:NSLayoutRelationEqual
                                                         toItem:self.tilesPerSquare[i-1]
                                                      attribute:NSLayoutAttributeWidth
                                                     multiplier:1
                                                       constant:0];
            [self addConstraint:constraint];
        }
        if (i>=64-8) {
            constraint = [NSLayoutConstraint constraintWithItem:self.tilesPerSquare[i]
                                                      attribute:NSLayoutAttributeTop
                                                      relatedBy:NSLayoutRelationEqual
                                                         toItem:self
                                                      attribute:NSLayoutAttributeTop
                                                     multiplier:1
                                                       constant:0];
            [self addConstraint:constraint];
        }
        if (i%8==7) {
            constraint = [NSLayoutConstraint constraintWithItem:self.tilesPerSquare[i]
                                                      attribute:NSLayoutAttributeRight
                                                      relatedBy:NSLayoutRelationEqual
                                                         toItem:self
                                                      attribute:NSLayoutAttributeRight
                                                     multiplier:1
                                                       constant:0];
            [self addConstraint:constraint];
        }
    }
    [self layoutIfNeeded];
}

#pragma mark - Other


- (void)reloadData
{
    UIView *squareOne = self.tilesPerSquare[0];

    for (int i=0; i<64; i++) {
        CGPoint point = CGPointMake((int)(i%8), (int)(i/8));
        FDChessboardPiece piece = [self.dataSource chessboardView:self
                                               pieceForCoordinate:point];
        self.piecesPerSquare[i] = @(piece);
        
        UIImageView *pieceImageView;
        NSString *file;
        UIColor *color;
        switch (piece) {
            case FDChessboardPieceEmpty:
                break;
            case FDChessboardPieceWhitePawn:
                file = @"wp";
                color = self.lightPieceColor;
                break;
            case FDChessboardPieceBlackPawn:
                file = @"bp";
                color = self.darkPieceColor;
                break;
            case FDChessboardPieceWhiteKnight:
                file = @"wn";
                color = self.lightPieceColor;
                break;
            case FDChessboardPieceBlackKnight:
                file = @"bn";
                color = self.darkPieceColor;
                break;
            case FDChessboardPieceWhiteBishop:
                file = @"wb";
                color = self.lightPieceColor;
                break;
            case FDChessboardPieceBlackBishop:
                file = @"bb";
                color = self.darkPieceColor;
                break;
            case FDChessboardPieceWhiteRook:
                file = @"wr";
                color = self.lightPieceColor;
                break;
            case FDChessboardPieceBlackRook:
                file = @"br";
                color = self.darkPieceColor;
                break;
            case FDChessboardPieceWhiteQueen:
                file = @"wq";
                color = self.lightPieceColor;
                break;
            case FDChessboardPieceBlackQueen:
                file = @"bq";
                color = self.darkPieceColor;
                break;
            case FDChessboardPieceWhiteKing:
                file = @"wk";
                color = self.lightPieceColor;
                break;
            case FDChessboardPieceBlackKing:
                file = @"bk";
                color = self.darkPieceColor;
                break;
        }
        if (file) {
            UIImage *image = [UIImage imageWithSVGNamed:file
                                             targetSize:CGSizeMake(200, 200)
                                              fillColor:color];
            pieceImageView = [[UIImageView alloc] initWithImage:image];
            pieceImageView.translatesAutoresizingMaskIntoConstraints = NO;
            self.pieceImageViewsPerSquare[i] = pieceImageView;
            [self addSubview:pieceImageView];
            NSLayoutConstraint *constraint;
            constraint = [NSLayoutConstraint constraintWithItem:pieceImageView
                                                      attribute:NSLayoutAttributeWidth
                                                      relatedBy:NSLayoutRelationEqual
                                                         toItem:squareOne
                                                      attribute:NSLayoutAttributeWidth
                                                     multiplier:1
                                                       constant:0];
            [self addConstraint:constraint];
            constraint = [NSLayoutConstraint constraintWithItem:pieceImageView
                                                      attribute:NSLayoutAttributeHeight
                                                      relatedBy:NSLayoutRelationEqual
                                                         toItem:squareOne
                                                      attribute:NSLayoutAttributeHeight
                                                     multiplier:1
                                                       constant:0];
            [self addConstraint:constraint];
            constraint = [NSLayoutConstraint constraintWithItem:pieceImageView
                                                      attribute:NSLayoutAttributeTop
                                                      relatedBy:NSLayoutRelationEqual
                                                         toItem:self.tilesPerSquare[i]
                                                      attribute:NSLayoutAttributeTop
                                                     multiplier:1
                                                       constant:0];
            [self addConstraint:constraint];
            constraint = [NSLayoutConstraint constraintWithItem:pieceImageView
                                                      attribute:NSLayoutAttributeLeading
                                                      relatedBy:NSLayoutRelationEqual
                                                         toItem:self.tilesPerSquare[i]
                                                      attribute:NSLayoutAttributeLeading
                                                     multiplier:1
                                                       constant:0];
            [self addConstraint:constraint];
        } else {
            self.pieceImageViewsPerSquare[i] = [NSNull null];
        }
    }
    [self layoutIfNeeded];
    
}

- (BOOL)movePieceAtCoordinate:(CGPoint)from toCoordinate:(CGPoint)to
{
    return YES;
}

- (BOOL)movePieceAtCoordinate:(CGPoint)from toCoordinate:(CGPoint)to andPromoteAs:(FDChessboardPiece)as
{
    return YES;
}

- (BOOL)premovePieceAtCoordinate:(CGPoint)from toCoordinate:(CGPoint)to
{
    return YES;
}

- (BOOL)premovePieceAtCoordinate:(CGPoint)from toCoordinate:(CGPoint)to andPromoteAs:(FDChessboardPiece)as
{
    return YES;
}


@end

File: Source/FDChessboardView.swift
//
//  FDChessboardView.swift
//  FDChessboardView
//
//  Created by William Entriken on 2/2/16.
//  Copyright © 2016 William Entriken. All rights reserved.
//

import Foundation
import UIKit

/// The location of a square on a chess board
public struct FDChessboardSquare: Hashable {
    /// From 0...63
    public let index: Int
    
    /// From 0...7 (a...h)
    public let file: Int
    
    /// From 0...7 (white king starting position to black king starting position)
    public let rank: Int
    
    /// Like a4
    public let algebraic: String

    public init(index newIndex: Int) {
        index = newIndex
        file = newIndex % 8
        rank = newIndex / 8
        algebraic = ["a","b","c","d","e","f","g","h"][file] + String(rank + 1)
    }
    
    public init(newRank: Int, newFile: Int) {
        rank = newRank
        file = newFile
        index = newRank * 8 + newFile
        algebraic = ["a","b","c","d","e","f","g","h"][file] + String(rank + 1)
    }
}

public protocol FDChessboardViewDataSource: class {
    /// What piece is on the square?
    func chessboardView(_ board: FDChessboardView, pieceForSquare square: FDChessboardSquare) -> FDChessboardPiece?

    /// The last move
    func chessboardViewLastMove(_ board: FDChessboardView) -> (from:FDChessboardSquare, to:FDChessboardSquare)?

    /// The premove
    func chessboardViewPremove(_ board: FDChessboardView) -> (from:FDChessboardSquare, to:FDChessboardSquare)?
}

public protocol FDChessboardViewDelegate: class {
    /// Where can this piece move to?
    func chessboardView(_ board: FDChessboardView, legalDestinationsForPieceAtSquare from: FDChessboardSquare) -> [FDChessboardSquare]

    /// After a move happened
    func chessboardView(_ board: FDChessboardView, didMoveFrom from: FDChessboardSquare, to: FDChessboardSquare, withPromotion promotion: FDChessboardPiece?)
}

/// Display for a chess board
@IBDesignable open class FDChessboardView: UIView {
    /// Color for "white" board squares
    @IBInspectable open var lightBackgroundColor: UIColor = UIColor(red: 222.0/255, green:196.0/255, blue:160.0/255, alpha:1)

    /// Color for "black" board squares
    @IBInspectable open var darkBackgroundColor: UIColor = UIColor(red: 160.0/255, green:120.0/255, blue:55.0/255, alpha:1)

    /// Color for where a piece is moving to
    open var targetBackgroundColor = UIColor(hue: 0.75, saturation:0.5, brightness:0.5, alpha:1.0)

    /// Color for a legal square target
    open var legalBackgroundColor = UIColor(hue: 0.25, saturation:0.5, brightness:0.5, alpha:1.0)

    /// Color for the last move square
    open var lastMoveColor = UIColor(hue: 0.35, saturation:0.5, brightness:0.5, alpha:1.0)

    /// Color for a premove square
    open var premoveColor = UIColor(hue: 0.15, saturation:0.5, brightness:0.5, alpha:1.0)

    /// Path for custom piece graphics
    open var pieceGraphicsDirectoryPath: String? = nil

    /// Datasource to say which pieces are on each square
    open weak var dataSource: FDChessboardViewDataSource? = nil {
        didSet {
            reloadData()
        }
    }

    /// Handler for user interaction with the view
    open weak var delegate: FDChessboardViewDelegate? = nil

    /// Should piece moves be animated?
    open var doesAnimate: Bool = true

    /// Should legal squares be shown when a piece is selected?
    open var doesShowLegalSquares: Bool = true

    /// Should the last move be shown?
    open var doesShowLastMove: Bool = true

    /// Should premove be shown?
    open var doesShowPremove: Bool = true

    private lazy var tilesAtIndex = [UIView]()

    private var pieceAtIndex = [Int : FDChessboardPiece]()

    private var pieceImageViewAtIndex = [Int : UIImageView]()

    private var lastMoveArrow: UIView? = nil

    private var premoveArrow: UIView? = nil

    /// UIView initializer
    public override init(frame: CGRect) {
        super.init(frame: frame)
        setup()
    }

    /// UIView initializer
    required public init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setup()
    }

    fileprivate func setup() {
        self.translatesAutoresizingMaskIntoConstraints = false

        for index in 0..<64 {
            let square = FDChessboardSquare(index: index)
            let tile = UIView()
            tile.backgroundColor = (index + index/8) % 2 == 0 ? darkBackgroundColor : lightBackgroundColor
            tile.translatesAutoresizingMaskIntoConstraints = false
            addSubview(tile)
            tilesAtIndex.append(tile)

            switch square.rank {
            case 0:
                addConstraint(NSLayoutConstraint(item: tile, attribute: .bottom, relatedBy: .equal, toItem: self, attribute: .bottom, multiplier: 1, constant: 0))
            case 7:
                addConstraint(NSLayoutConstraint(item: tile, attribute: .bottom, relatedBy: .equal, toItem: tilesAtIndex[index - 8], attribute: .top, multiplier: 1, constant: 0))
                addConstraint(NSLayoutConstraint(item: tile, attribute: .top, relatedBy: .equal, toItem: self, attribute: .top, multiplier: 1, constant: 0))
                addConstraint(NSLayoutConstraint(item: tile, attribute: .height, relatedBy: .equal, toItem: tilesAtIndex[index - 8], attribute: .height, multiplier: 1, constant: 0))
            default:
                addConstraint(NSLayoutConstraint(item: tile, attribute: .bottom, relatedBy: .equal, toItem: tilesAtIndex[index - 8], attribute: .top, multiplier: 1, constant: 0))
                addConstraint(NSLayoutConstraint(item: tile, attribute: .height, relatedBy: .equal, toItem: tilesAtIndex[index - 8], attribute: .height, multiplier: 1, constant: 0))
            }
            switch square.file {
            case 0:
                addConstraint(NSLayoutConstraint(item: tile, attribute: .left, relatedBy: .equal, toItem: self, attribute: .left, multiplier: 1, constant: 0))
            case 7:
                addConstraint(NSLayoutConstraint(item: tile, attribute: .left, relatedBy: .equal, toItem: tilesAtIndex[index - 1], attribute: .right, multiplier: 1, constant: 0))
                addConstraint(NSLayoutConstraint(item: tile, attribute: .right, relatedBy: .equal, toItem: self, attribute: .right, multiplier: 1, constant: 0))
                addConstraint(NSLayoutConstraint(item: tile, attribute: .width, relatedBy: .equal, toItem: tilesAtIndex[index - 1], attribute: .width, multiplier: 1, constant: 0))
            default:
                addConstraint(NSLayoutConstraint(item: tile, attribute: .left, relatedBy: .equal, toItem: tilesAtIndex[index - 1], attribute: .right, multiplier: 1, constant: 0))
                addConstraint(NSLayoutConstraint(item: tile, attribute: .width, relatedBy: .equal, toItem: tilesAtIndex[index - 1], attribute: .width, multiplier: 1, constant: 0))
            }
        }
        self.layoutIfNeeded()
    }

    /*
- (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event
{
UITouch *touch = [touches anyObject];
CGPoint point = [touch locationInView:self];
NSInteger x = point.x*8/self.frame.size.width;
NSInteger y = 8-point.y*8/self.frame.size.height;
UIView *tile = self.tilesPerSquare[y*8+x];
tile.backgroundColor = self.targetBackgroundColor;
}

- (void)touchesMoved:(NSSet *)touches withEvent:(UIEvent *)event
{
UITouch *touch = [touches anyObject];
CGPoint point = [touch locationInView:self];
if (![self pointInside:point withEvent:event])
return;
NSInteger x = point.x*8/self.frame.size.width;
NSInteger y = 8-point.y*8/self.frame.size.height;
UIView *tile = self.tilesPerSquare[y*8+x];
tile.backgroundColor = self.legalBackgroundColor;
}
*/

    /// Add a piece onto the board
    open func setPiece(_ piece: FDChessboardPiece?, forSquare square: FDChessboardSquare){
        let index = square.index
        pieceAtIndex[index] = piece
        self.pieceImageViewAtIndex[index]?.removeFromSuperview()
        self.pieceImageViewAtIndex.removeValue(forKey: index)
        guard let piece = piece else {
            return
        }

        let pieceImageView: UIImageView
        let fileName = piece.rawValue
        let bundle = Bundle(for: type(of: self))
        let image = UIImage(named: fileName, in: bundle, compatibleWith: nil)
        pieceImageView = UIImageView(image: image)
        pieceImageView.translatesAutoresizingMaskIntoConstraints = false
        self.pieceImageViewAtIndex[index] = pieceImageView
        self.addSubview(pieceImageView)

        let squareOne = self.tilesAtIndex.first!
        self.addConstraint(NSLayoutConstraint(item: pieceImageView, attribute: .width, relatedBy: .equal, toItem: squareOne, attribute: .width, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint(item: pieceImageView, attribute: .height, relatedBy: .equal, toItem: squareOne, attribute: .height, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint(item: pieceImageView, attribute: .top, relatedBy: .equal, toItem: self.tilesAtIndex[index], attribute: .top, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint(item: pieceImageView, attribute: .leading, relatedBy: .equal, toItem: self.tilesAtIndex[index], attribute: .leading, multiplier: 1, constant: 0))
        self.layoutIfNeeded()
    }

    /// Repull all board information from data source
    open func reloadData() {
        for index in 0 ..< 64 {
            let square = FDChessboardSquare(index: index)
            let newPiece = dataSource?.chessboardView(self, pieceForSquare:square)
            setPiece(newPiece, forSquare: square)
        }
        self.layoutIfNeeded()
    }

    /// Move a piece on the board
    open func movePieceAtCoordinate(_ from: FDChessboardSquare, toCoordinate to: FDChessboardSquare) -> Bool {
        return true
    }

    /// Move a piece on the board
    open func movePieceAtCoordinate(_ from: FDChessboardSquare, toCoordinate to: FDChessboardSquare, andPromoteTo piece: FDChessboardPiece) -> Bool {
        return true
    }

    /// Premove a piece on the board
    open func premovePieceAtCoordinate(_ from: FDChessboardSquare, toCoordinate to: FDChessboardSquare) -> Bool {
        return true
    }

    /// Premove a piece on the board
    open func premovePieceAtCoordinate(_ from: FDChessboardSquare, toCoordinate to: FDChessboardSquare, andPromoteTo piece: FDChessboardPiece) -> Bool {
        return true
    }

}
